apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -0,0 +1,423 @@
+    end.
+
+%%================================================================================
+%% Internal functions
+%%================================================================================
+
+next_loop(_ITHandle, _KeyMapper, _Filter, _Cutoff, It, Acc, 0) ->
+    {ok, It, lists:reverse(Acc)};
+next_loop(ITHandle, KeyMapper, Filter, Cutoff, It0, Acc0, N0) ->
+    inc_counter(),
+    #it{last_seen_key = Key0} = It0,
+    case emqx_ds_bitmask_keymapper:bin_increment(Filter, Key0) of
+        overflow ->
+            {ok, It0, lists:reverse(Acc0)};
+        Key1 ->
+            %% assert
+            true = Key1 > Key0,
+            case rocksdb:iterator_move(ITHandle, {seek, Key1}) of
+                {ok, Key, Val} ->
+                    {N, It, Acc} =
+                        traverse_interval(ITHandle, Filter, Cutoff, Key, Val, It0, Acc0, N0),
+                    next_loop(ITHandle, KeyMapper, Filter, Cutoff, It, Acc, N);
+                {error, invalid_iterator} ->
+                    {ok, It0, lists:reverse(Acc0)}
+            end
+    end.
+
+traverse_interval(ITHandle, Filter, Cutoff, Key, Val, It0, Acc0, N) ->
+    It = It0#it{last_seen_key = Key},
+    case emqx_ds_bitmask_keymapper:bin_checkmask(Filter, Key) of
+        true ->
+            Msg = deserialize(Val),
+            case check_message(Cutoff, It, Msg) of
+                true ->
+                    Acc = [Msg | Acc0],
+                    traverse_interval(ITHandle, Filter, Cutoff, It, Acc, N - 1);
+                false ->
+                    traverse_interval(ITHandle, Filter, Cutoff, It, Acc0, N);
+                overflow ->
+                    {0, It0, Acc0}
+            end;
+        false ->
+            {N, It, Acc0}
+    end.
+
+traverse_interval(_ITHandle, _Filter, _Cutoff, It, Acc, 0) ->
+    {0, It, Acc};
+traverse_interval(ITHandle, Filter, Cutoff, It, Acc, N) ->
+    inc_counter(),
+    case rocksdb:iterator_move(ITHandle, next) of
+        {ok, Key, Val} ->
+            traverse_interval(ITHandle, Filter, Cutoff, Key, Val, It, Acc, N);
+        {error, invalid_iterator} ->
+            {0, It, Acc}
+    end.
+
+-spec check_message(emqx_ds:time(), iterator(), emqx_types:message()) ->
+    true | false | overflow.
+check_message(
+    Cutoff,
+    _It,
+    #message{timestamp = Timestamp}
+) when Timestamp >= Cutoff ->
+    %% We hit the current epoch, we can't continue iterating over it yet.
+    %% It would be unsafe otherwise: messages can be stored in the current epoch
+    %% concurrently with iterating over it. They can end up earlier (in the iteration
+    %% order) due to the nature of keymapping, potentially causing us to miss them.
+    overflow;
+check_message(
+    _Cutoff,
+    #it{start_time = StartTime, topic_filter = TopicFilter},
+    #message{timestamp = Timestamp, topic = Topic}
+) when Timestamp >= StartTime ->
+    emqx_topic:match(emqx_topic:words(Topic), TopicFilter);
+check_message(_Cutoff, _It, _Msg) ->
+    false.
```

Comment:
Q: wouldn't `bin_increment` with the `{StartTime, '..', SafeCutoffTime - 1}` constraint on timestamps prevent `Timestamp < StartTime`?

Reply:
I kinda like the explicit code here. @keynslug What do you thing?

Reply:
Currently, as per my understanding, there are situations when constraints in `emqx_ds_bitmask_keymapper` are not followed strictly (see [this testcase](https://github.com/ieQu1/emqx/blob/a1cdbaa76d1dc4f7a9e1287f8ba09300512e6001/apps/emqx_durable_storage/src/emqx_ds_bitmask_keymapper.erl#L754-L759) for example). So, these are currently more like hints rather than requirements, therefore it seems necessary to put some guardrails on this level.